### ISSUE TITLE
RFC: make tarballs first class citizen for Kata, and fix OBS shortcomings?

### ISSUE THREAD

**Issue by egernst** [kata-containers/runtime#1080]
# Description of problem

I started to open issues in documentation and packaging repositories, but perhaps we should have a more general conversation on the topic.

# Current gaps with OBS
I have been uncomfortable with how we handle and provide packages for Kata Containers.  Today, each branch (stable latest, stable older and master) has its own OBS sub-project.  For a user this means each branch is its own separate packaging URL.  So, if a user wants to move from stable to master, they'll need to update their packaging managers' URLs.

A second issue I see is that we only offer the latest package.  You cannot install the prior master or stable release.  This isn't great from a user perspective.

Third, as OSVs add support for Kata, we are duplicating efforts in our heavy and sometimes error prone packaging management.  

# tarballs?

I see many other projects make use of tarballs for releases today.  With Kata Containers, we already have these available as part of the release process, and they are the foundation of cloud-native methods for installing Kata (see the kata-deploy tooling in the packaging repo), [here](https://github.com/kata-containers/packaging/blob/master/kata-deploy/README.md).

With static binaries we already provide, we can easily make use of github to be the source of our artifacts, including prior releases.  As an example, it was very easy to get users started on 'whatever distro' with Kata with firecracker:

```
wget https://github.com/kata-containers/runtime/releases/download/1.5.0-rc2/kata-fc-static-1.5.0-rc2-x86_64.tar.gz
sudo tar -xvf kata-fc-static-1.5.0-rc2-x86_64.tar.gz -C /
```

When used with something like kata-deploy, post-day-0 is handled pretty well (updated through the daemonset).  I think the main drawback/consideration will be difference in behavior an end user in production (ie, post-day-0 - how to handle updates effectively).  But, in cloud-native, I think this would be acceptable anyway?
 
I'd like to get feedback from others on this.   

I'd *really* like to see this be a (or, the) recommended method for getting started with Kata, and tested in our CI.

**Comment by bryteise** [comment]
@egernst I expect most of your users to want to grab snapshot binaries and release tarballs with static content is definitely appreciated in that case. The OBS use case would really be for a distro where you also include dependencies to avoid distro dependencies breaking (also enable easier to run validation). That is a significant amount of effort though and I'm not sure how valuable it is for your end users.

For distros who may eventually pick up kata and build it themselves, just having a good build system story is all you should really need. Aim to make the build work offline and provide all source dependencies as part of your release as that's kind of how things are best done in the go world at the moment.

**Comment by gnawux** [comment]
I agree with the tarball idea. But could we make real static binaries, especially for the networking related binaries?

**Comment by jodh-intel** [comment]
# Notes

> So, if a user wants to move from stable to master, they'll need to update their packaging managers' URLs.

We could provide something like an `update-alternatives(1)` command that would allow the URL to be switched easily. But maybe something like https://packagecloud.io (coupled with OBS?) would work as that provides a standard repo layout fwics (for example, https://packagecloud.io/AtomEditor/atom includes multiple versions of atom).

> You cannot install the prior master or stable release. This isn't great from a user perspective.

You can do this (in principle) using snap packages.

> Third, as OSVs add support for Kata, we are duplicating efforts in our heavy and sometimes error prone packaging management.

I agree that the ideal would be for OSV's to package Kata natively as that would remove the need for us to create packages. And if they do that, users would be able to rely on the particular distro-specific infrastructure to potentially switch from a "current" package version to, say, an LTS one.

# More info

I'm not convinced tarballs are the way forward for everyone. They are simple, but I don't think we want to recommend users start running commands like that `sudo tar xvf $file /` on their primary system.

## Pros and cons

Pros of tarballs:

- They make cloud deployment (slightly?) easier.

   I'd like a bit more info on how/why this is tbh.

- ?

Cons of tarballs:

- `tar` syntax is confusing (and therefore potentially dangerous) for naive users.
- No built-in security (no gpg sigs, etc).
- No automatic updates (potential security concerns).
  In fact, if cloud folk want to rely on tarballs, wouldn't they need a machine-readable method to determine new releases so that new tarballs can be sucked down and installed using their preferred cloud infrastructure? Currently, this could be achieved by using the github api to look for new releases I guess but is this the best way?

Pros of using distro-native tooling:

- Familiar to users.
- Safer to run.
- Security built-in.
- Already documented.
- Dependency-resolution.
- Automatic updates.
- Easy for admins to manage.

Cons of using distro-native tooling:

- Different distros use different commands (`apt-get`, `dnf`, *etc*).
- ?

# TL;DR

There are multiple issues here which need to be dealt with individually:

- OBS limitations.
- The suggestion that we switch from distro-native packaging to simple tarballs.
- OSV packaging efforts.
- Provision of different build types: static binaries versus dynamic.
- Differing classes of users: basic user, power user, developer, cloud, snap user, OSV, distro end-user, etc.

Overall, this sounds like a good topic for discussion at the Architecture Committee meeting.

**Comment by mcastelino** [comment]
> For distros who may eventually pick up kata and build it themselves, just having a good build system story is all you should really need. Aim to make the build work offline and provide all source dependencies as part of your release as that's kind of how things are best done in the go world at the moment.

@bryteise you are right. And given that clearlinux always bundles kata and its dependencies, can we not assume that if kata can be built on Clearlinux that requirement is satisified and can be used as a reference for other distributions.
